# Data Analysis: Profitable App Profiles (App Store and Google Play Markets)

What we want from this project is to use data to find what kinds of apps from Google Play and the App Store attract more users and make more profit. We are working as data analysts for a company that builds Android and iOS mobile apps.

Our company only builds apps that are free to download and install, and our main source of revenue comes from in-app ads. In other words, our revenue is mostly influenced by the number of users that use our apps. For this project, our goal is to analyze data to help our team of developers understand the types of apps that attract more users and are hence more profitable.

## Opening and Exploring the Data

In [1]:
from csv import reader

# Google Play Data Set
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

# Apple Store Data Set
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

Below is a function, written by Dataquest, named `explore_data()` that can be used repeatedly to print rows in a readable way. We can also print the number of row and columns if the argument passed to the parameter `rows_and_columns` is `True`.

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
print(android_header)
print('\n')
explore_data(android, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


The Google Play data set includes 10841 apps and 13 columns. Some columns that might help us with our analysis are: `'App'`, `'Category'`, `'Reviews'`, `'Installs'`, `'Type'`, `'Price'`, and `'Genres'`.

For better descriptions of each of Google Play data set's column names, use this [documentation](https://www.kaggle.com/lava18/google-play-store-apps).

In [4]:
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


The App Store data set includes 7197 apps and 16 columns. Some columns that might help us with our analysis are: `'track_name'`, `'currency'`, `'price'`, `'rating_count_tot'`, `'rating_count_ver'`, and `'prime_genre'`.

For better descriptions of each of the App Store data set's column names, use this [documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps).

## Deleting Wrong Data

The Google Play data set has a dedicated [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), and [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) describes an error for row 10472. We will print the row and check it against the header and another row.

In [5]:
print(android[10472])
print('\n')
print(android_header)
print('\n')
print(android[0])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


When checking row 10472 (the *Life Made WI-Fi Touchscreen Photo Frame* app) against the header, we see a rating of 19 as well as a price of 'Everyone'. Ratings are out of 5 and prices are numerical, so neither makes sense. Hence, we'll delete row 10472 and check the number of entries in the data set before and after.

In [6]:
print(len(android))
del android[10472]
print(len(android))

10841
10840


From a glance at the App Store data set's [discussion section](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion), we do not see any reports of wrong data.

## Removing Duplicate Entries

### Part One

After further exploring the Google Play data set's [discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion) section, we see that some apps have duplicate entries. For example, *Instagram* has four entries:

In [7]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Let's count the number of duplicates we have in the Google Play data set:

In [8]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:10])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


There are a total of 1181 cases where an app is repeated in the Google Play data set.

Since we do not want any one app to be counted more than once when we do our data analysis, we will remove duplicate entries, keeping only one entry per app. Although we could remove the duplicates randomly, there are better ways at deciding which entry to keep.

In the *Instagram* app case, the only difference between each entry is the fourth position of each row, which corresponds to the number of reviews. If we assume the different numbers show the data was collected at different times, we can choose to keep the row with the highest number of reviews since it should be the more recent data, and remove the other entries.

To remove the duplicates, we will:
- Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
- Use the dictionary to create a new data set, which will have only one entry per app (we'll only select the entry with the highest number of reviews)

### Part Two

We will first start by creating the dictionary.

In [9]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

Since we previously found that there are 1181 cases of duplicate entries, after removing the duplicates, we should expect there to be $10840 - 1181 = 9659$ entries.

In [10]:
print('Expected length:', len(android) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


Using the `reviews_max` dictionary we created, we will remove duplicates, keeping only the entries with the highest number of reviews. To do this, in the cell below we will:
- Create two empty lists: `android_clean` and `already_added`.
- Loop through Google Play data set, and for each iteration:
    - Assign app name to the variable `name`, and convert number of reviews to `float` then assign to the variable `n_reviews`.
    - Add the current row (`app`) to the `android_clean` list and the app name (`name`) to the `already_added` list IF:
        - The number of reviews of the current app matches the number of reviews of the same app in the `reviews_max` dictionary; AND
        - The app name `name` is not already in the `already_added` list. This condition is so that we do not add the same app more than once for the cases where the highest number of reviews of duplicates is the same for more than one entry.

In [11]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)

We will now explore the `android_clean` data set and ensure it has the expected 9659 rows.

In [12]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


We have confirmed that there are indeed 9659 rows, just like we expected.

Upon checking the App Store data set's [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), there does not appear to be any duplicate entries in the App Store data set.

## Removing Non-English Apps

### Part One

We use English for the apps we develop, and we'd like to analyze only the apps directed toward an English-speaking audience. Since we are not interested in keeping non-English apps, we'll remove them.

We will remove each app with a name containing a symbol not commonly used in English text, where English text usually consists of letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, \*, /, etc).

According to the ASCII system, the numbers corresponding to characters commonly used in English text range from 0 to 127. We can build a function using the built-in `ord()` function to detect if an app name contains a character greater than 127.

In [13]:
def is_english(string):
    for character in string:
        if ord(character) > 127:
            return False
    return True

We will now test this function to see if it correctly detects wheher an app name is English or non-English:

In [14]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


The function seems to work correctly for detecting app names with non-English characters, but it also seems to categorize English app names as non-English if the app name includes emojis or other symbols like ™.

In [15]:
print(ord('™'))
print(ord('😜'))

8482
128540


We do not want to remove English apps just because their app names are incorrectly detected as non-English due to including emojis or other symbols (like ™) outside the ASCII range. This would mess with our analysis of English apps.

### Part Two

To minimize the impact of data loss, we'll only remove an app if its name has more than three characters outside our ASCII range:

In [16]:
def is_english(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
            
    if non_ascii > 3:
        return False
    else:    
        return True

In [17]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Our filter is still not perfect, but it should be effective enough for our analysis.

We will use our `is_english()` function to filter out non-English apps from both data sets:

In [18]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)
        
for app in ios:
    name = app[1]
    if is_english(name):
        ios_english.append(app)

In [19]:
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

We now have 9614 Android apps and 6183 iOS apps remaining.

## Isolating the Free Apps
In the intro, we mentioned that we only build apps that are free to download and install, and our main source of revenue comes from in-app ads. Since our data sets contain both free and non-free apps, we will need to isolate only the free apps for our analysis.

In [20]:
# Identify columns describing the app price
print(android_header)
print('\n')
print(android[0])
print('\n')
print(ios_header)
print('\n')
print(ios[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


In [21]:
android_free = []
ios_free = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_free.append(app)

for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_free.append(app)

android_final = android_free
ios_final = ios_free

# Check length of each data set
# How many apps remaining?
print(len(android_final))
print(len(ios_final))

8864
3222


We have 8864 Android apps and 3222 iOS apps remaining. This should be enough for us to proceed with our analysis.

## Most Common Apps by Genre
### Part One

As mentioned in the intro, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:
1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Our end goal is to add the app on both Google Play and the App Store, so we need to find app profiles successful on both markets. For example, a profile that works well for both markets might be a productivity app that makes use of gamification.

We'll begin the analysis by getting a sense of what are the most common genres for each market. To do this, we will build a frequency table for `'Category'` and `'Genres'` columns of the Google Play data set and the `'prime_genre'` column of the App Store data set.

### Part Two

We will build two functions we can use to analyze the frequency tables:
- One function to generate frequency tables that show percentages
- Another function we can use to display the percentages in a descending order

In [22]:
def freq_table(dataset, index):
    table = {}
    total_count = 0
    
    # Create frequency table as dictionary
    for row in dataset:
        total_count += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    # Express frequencies as percentages
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total_count) * 100
        table_percentages[key] = percentage
        
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

### Part Three

We will start by analyzing the frequency table for the `'prime_genre'` column of the App Store data set.

In [23]:
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Among the free English apps, we can see that the most common type of app is for games (58.16%), followed by entertainment (7.88%), then photo and video (4.97%), education (3.66%), and social networking (3.29%).

The general impression is that most of the App Store's free English apps are designed for entertainment (games, photo and video, social networking, sports, music, etc.) rather than for practical purposes (education, shopping, utilities, productivity, lifestyle, etc.). However, a larger number of entertainment apps does not also imply that entertainment apps generally have a larger number of users.

We will next analyze the frequency table for the `'Category'` and `'Genres'` columns of the Google Play data set.

In [24]:
# Category column
display_table(android_final, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

Unlike the free English apps in the App Store where many of their apps were targeted toward entertainment, among the free English apps in Google Play, most of the apps (by category) seem to focus on practical purposes (family, tools, business, lifestyle, productivity, finance, etc.). The most common category of free English apps in Google Play is family (18.91%), followed by games (9.72%), tools (8.46%), business (4.59%), lifestyle (3.90%), and so forth.

In [25]:
# Genres column
display_table(android_final, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

Similar to what we saw when analyzing the frequency table for category, we can see that when analyzing genres, the majority of Google Play's free English apps are still for practical purposes instead of entertainment. The most common genres are as follows: tools (8.45%), entertainment (6.07%), education (5.35%), business (4.59%), productivity (3.89%), lifestyle (3.89%), and so on.

While we are not exactly sure how the `Category` and `Genres` columns are distinguished, `Genres` definitely has more categories. Further analysis for Google Play will be based on the `Category` column.

Although this still does not reveal which genres have the most users, we can definitely see that the App Store has over half its store covered in entertainment (i.e, games) apps while Google Play displays a better mixture of practical purposes to entertainment apps.

## Most Popular Apps by Genre on the App Store

So far, we've discovered what kinds of apps dominate the App Store and the Google Play market. Now, we want to get an idea of what kind of apps have the most users.

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we'll use the `Installs` column. The App Store data set lacks a column with information on number of installs, so we'll settle for the total number of user ratings under the `rating_count_tot` column.

We'll start by calculating the average number of user ratings for each app genre on the App Store:

In [26]:
genres_ios = freq_table(ios_final, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Health & Fitness : 23298.015384615384
Catalogs : 4004.0
Finance : 31467.944444444445
Navigation : 86090.33333333333
Entertainment : 14029.830708661417
Weather : 52279.892857142855
Sports : 23008.898550724636
Music : 57326.530303030304
Business : 7491.117647058823
Photo & Video : 28441.54375
Utilities : 18684.456790123455
Productivity : 21028.410714285714
Medical : 612.0
News : 21248.023255813954
Shopping : 26919.690476190477
Games : 22788.6696905016
Book : 39758.5
Education : 7003.983050847458
Social Networking : 71548.34905660378
Travel : 28243.8
Food & Drink : 33333.92307692308
Reference : 74942.11111111111
Lifestyle : 16485.764705882353


On average, navigation apps have the highest number of user reviews, followed by reference apps, then social networking apps, then music apps, then weather apps, book apps, food and drink apps, and so forth. 

Let's see what apps some of these genres consist of:

In [27]:
for app in ios_final:
    if app[-5] == 'Navigation':
        print(app[1], ":", app[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The average number of user ratings for navigation apps seems to be heavily influenced by Waze and Google Maps. The two apps alone account for $345046+154911=499957$ (nearly half a million) user reviews.

In [28]:
for app in ios_final:
    if app[-5] == 'Reference':
        print(app[1], ":", app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


As you can see, although apps of the reference genre are among the second highest average number of reviews, this is heavily influenced by the Bible and Dictionary.com and their combined total of over a million reviews.

In [29]:
for app in ios_final:
    if app[-5] == 'Social Networking':
        print(app[1], ":", app[5])

Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie Talkie : 27662
Miitomo : 2

Similar to how just a couple of popular apps inflate the average number of user ratings for the navation genre and the reference genre, the average number for social networking apps seems to also be heavily influenced by a few of the more popular apps like Facebook, Pinterest, Skype, Messenger, etc.

In [30]:
for app in ios_final:
    if app[-5] == 'Music':
        print(app[1], ":", app[5])

Pandora - Music & Radio : 1126879
Spotify Music : 878563
Shazam - Discover music, artists, videos & lyrics : 402925
iHeartRadio – Free Music & Radio Stations : 293228
SoundCloud - Music & Audio : 135744
Magic Piano by Smule : 131695
Smule Sing! : 119316
TuneIn Radio - MLB NBA Audiobooks Podcasts Music : 110420
Amazon Music : 106235
SoundHound Song Search & Music Player : 82602
Sonos Controller : 48905
Bandsintown Concerts : 30845
Karaoke - Sing Karaoke, Unlimited Songs! : 28606
My Mixtapez Music : 26286
Sing Karaoke Songs Unlimited with StarMaker : 26227
Ringtones for iPhone & Ringtone Maker : 25403
Musi - Unlimited Music For YouTube : 25193
AutoRap by Smule : 18202
Spinrilla - Mixtapes For Free : 15053
Napster - Top Music & Radio : 14268
edjing Mix:DJ turntable to remix and scratch music : 13580
Free Music - MP3 Streamer & Playlist Manager Pro : 13443
Free Piano app by Yokee : 13016
Google Play Music : 10118
Certified Mixtapes - Hip Hop Albums & Mixtapes : 9975
TIDAL : 7398
YouTube Mu

Music apps are in the same situation where a few popular apps (Pandora, Spotify, Shazam, etc.) heavily influence the average number.

We are trying to find popular genres, but due to a few apps that have as much as over ten times the number of reviews as the other apps within the same genre, these said genres might appear to be much more popular than they really are.

Nonetheless, it does appear that entertainment apps dominate the App Store. Based on our findings, an app that could possibly do well on the App Store could be a navigation tool like Google Maps, but with built in music stations like Spotify, so someone on the road won't have to switch back and forth between two apps when listening to music while navigating.

## Most Popular Apps by Genre on Google Play

For the Google Play market, we have data about the number of installs (unlike for the App Store), but the numbers are not too precise. In fact, the values are mostly open-ended (100+, 1,000+, 5,000+, etc.):

In [31]:
# Installs columns
display_table(android_final, 5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


However, since we only want to find out which apps attract the most users, we don't need very precise data. We will assume apps with 100+ installs have 100 installs, apps with 100,000+ installs have 100,000 installs, apps with 1,000,000+ installs have 1,000,000 installs, and so on.

In order to carry out computations, we will need to convert each install number from `string` to `float`. We will then compute the average number of installs for each app genre (or in our case, category).

In [32]:
categories_android = freq_table(android_final, 1)

for category in categories_android:
    total = 0
    len_category = 0
    
    for app in android_final:
        category_app = app[1]
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace('+', '')
            n_installs = n_installs.replace(',', '')
            total += float(n_installs)
            len_category += 1
    
    avg_n_installs = total / len_category
    print(category, ":", avg_n_installs)

BEAUTY : 513151.88679245283
HOUSE_AND_HOME : 1331540.5616438356
PARENTING : 542603.6206896552
EDUCATION : 1833495.145631068
PRODUCTIVITY : 16787331.344927534
COMMUNICATION : 38456119.167247385
WEATHER : 5074486.197183099
SPORTS : 3638640.1428571427
EVENTS : 253542.22222222222
ENTERTAINMENT : 11640705.88235294
FINANCE : 1387692.475609756
BOOKS_AND_REFERENCE : 8767811.894736841
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
PHOTOGRAPHY : 17840110.40229885
LIBRARIES_AND_DEMO : 638503.734939759
FAMILY : 3695641.8198090694
MAPS_AND_NAVIGATION : 4056941.7741935486
BUSINESS : 1712290.1474201474
GAME : 15588015.603248259
DATING : 854028.8303030303
COMICS : 817657.2727272727
MEDICAL : 120550.61980830671
AUTO_AND_VEHICLES : 647317.8170731707
PERSONALIZATION : 5201482.6122448975
SHOPPING : 7036877.311557789
LIFESTYLE : 1437816.2687861272
TRAVEL_AND_LOCAL : 13984077.710144928
NEWS_AND_MAGAZINES : 9549178.467741935
VIDEO_PLAYERS : 24727872.452830188
SOCIAL : 23253652.127118

On average, it appears communication apps have the most installs: 38,456,119. This number, similar to what we saw with the App Store's app genres, might be heavily influenced by a select handful of more popular apps.

Let's see which apps have over 100 million, 500 million, or 1 billion installs:

In [34]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '100,000,000+'
                                     or app[5] == '500,000,000+'
                                     or app[5] == '1,000,000,000+'):
        print(app[0], ":", app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

As we can see, the number of installs for communication apps are heavily skewed by these apps, especially the apps with a billion or more installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts).

Let's see the average number of installs for communications apps if we remove all the ones with over 100 million installs:

In [36]:
under100mil = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace('+', '')
    n_installs = n_installs.replace(',', '')
    
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under100mil.append(float(n_installs))
        
sum(under100mil) / len(under100mil)

3603485.3884615386

The original average number of installs for communication apps (38,456,119) is reduced to 3,603,485 after removing the communication apps with over 100 million installs. The average was reduced to less than ten percent of the original!

If the patterns are anything like those of the App Store, we can assume a similar occurrence of app catergories with the higher average installs being heavily influenced by a few more popular apps within the categories for video players apps (24,727,872), social apps (23,253,652), photography apps (17,840,110), and productivity apps (16,787,331). As for games (15,588,015), the high average is just due to there being a high proportion of game apps to the total number of all apps.

The next category for highest average number of installs is the travel and local category (13,984,077). Since our goal is to recommend an app genre/category with potential of being profitable on both the App Store and Google Play, and since we previously saw that navigation with music has possible potential, let's explore some of the travel and local apps as well as their number of installs.

In [37]:
for app in android_final:
    if app[1] == 'TRAVEL_AND_LOCAL':
        print(app[0], ":", app[5])

trivago: Hotels & Travel : 50,000,000+
Hopper - Watch & Book Flights : 5,000,000+
TripIt: Travel Organizer : 1,000,000+
Trip by Skyscanner - City & Travel Guide : 500,000+
CityMaps2Go Plan Trips Travel Guide Offline Maps : 1,000,000+
KAYAK Flights, Hotels & Cars : 10,000,000+
World Travel Guide by Triposo : 500,000+
Booking.com Travel Deals : 100,000,000+
Hostelworld: Hostels & Cheap Hotels Travel App : 1,000,000+
Google Trips - Travel Planner : 5,000,000+
GPS Map Free : 5,000,000+
GasBuddy: Find Cheap Gas : 10,000,000+
Southwest Airlines : 5,000,000+
AT&T Navigator: Maps, Traffic : 10,000,000+
VZ Navigator : 50,000,000+
KakaoMap - Map / Navigation : 10,000,000+
AirAsia : 10,000,000+
Expedia Hotels, Flights & Car Rental Travel Deals : 10,000,000+
Goibibo - Flight Hotel Bus Car IRCTC Booking App : 10,000,000+
Allegiant : 1,000,000+
Amtrak : 1,000,000+
JAL (Domestic and international flights) : 1,000,000+
Flight & Hotel Booking App - ixigo : 5,000,000+
VZ Navigator for Tablets : 500,000+

Let's see how many of these apps could be possibly skewing our average number of installs:

In [38]:
for app in android_final:
    if app[1] == 'TRAVEL_AND_LOCAL' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Booking.com Travel Deals : 100,000,000+
TripAdvisor Hotels Flights Restaurants Attractions : 100,000,000+
Maps - Navigate & Explore : 1,000,000,000+
Google Street View : 1,000,000,000+
Google Earth : 100,000,000+


As we can see, there are only 5 apps of the travel and local category that are highly popular with over 100 million installs. Let's see which apps are decently popular (1 million to 100 million installs):

In [39]:
for app in android_final:
    if app[1] == 'TRAVEL_AND_LOCAL' and (app[5] == '1,000,000+'
                                        or app[5] == '5,000,000+'
                                        or app[5] == '10,000,000+'
                                        or app[5] == '50,000,000+'):
        print(app[0], ":", app[5])

trivago: Hotels & Travel : 50,000,000+
Hopper - Watch & Book Flights : 5,000,000+
TripIt: Travel Organizer : 1,000,000+
CityMaps2Go Plan Trips Travel Guide Offline Maps : 1,000,000+
KAYAK Flights, Hotels & Cars : 10,000,000+
Hostelworld: Hostels & Cheap Hotels Travel App : 1,000,000+
Google Trips - Travel Planner : 5,000,000+
GPS Map Free : 5,000,000+
GasBuddy: Find Cheap Gas : 10,000,000+
Southwest Airlines : 5,000,000+
AT&T Navigator: Maps, Traffic : 10,000,000+
VZ Navigator : 50,000,000+
KakaoMap - Map / Navigation : 10,000,000+
AirAsia : 10,000,000+
Expedia Hotels, Flights & Car Rental Travel Deals : 10,000,000+
Goibibo - Flight Hotel Bus Car IRCTC Booking App : 10,000,000+
Allegiant : 1,000,000+
Amtrak : 1,000,000+
JAL (Domestic and international flights) : 1,000,000+
Flight & Hotel Booking App - ixigo : 5,000,000+
Wisepilot for XPERIA™ : 5,000,000+
VZ Navigator for Galaxy S4 : 5,000,000+
MAIN : 1,000,000+
Yoriza Pension - travel, lodging, pension, camping, caravan, pool villas ac

Among the travel and local apps with between 1 million to 100 million installs, there appears to many apps for booking hotels, booking flights, and for navigating. These will be our competition if we decide to build an app too similar.

There are already plenty of navigators available on the market, so to make ours stand out, we would need to add features that are, let's say, roadtrip friendly. We can add a music feature such as an incorporated radio station. There could be a special tab that lists popular music artists in the area you're traveling in so you could switch to those artists' radio stations and immerse yourself in your destination's culture. There could also be a tab for most recommended restaurants in the area, closest gas stations or public restrooms, etc.

## Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps (specifically English apps) with the goal of recommending an app profile that can be profitable for both markets.

We concluded that making a navigation app with special features could be profitable for both the App Store and Google Play markets. Since the markets already have plenty of navigation apps, we would need to make ours different and more desirable. This could include adding a music feature that lists popular music in your travel destination, so you could switch to that music's radio stations (in the navigation app) and immerse yourself in your destination's culture. There could also be a feature that displays most recommended restaurants in the area, closest gas stations, cleanest (as rated by public app users) public restrooms, etc.